server/data/restClientBuilder.ts

```diff
@@ -0,0 +1,137 @@
+import superagent from 'superagent'
+import Agent, { HttpsAgent } from 'agentkeepalive'
+import logger from '../../log'
+import { buildErrorHandler } from './clientErrorHandler'
+
+interface TimeoutSpec {
+  response: number
+  deadline: number
+}
+
+interface AgentConfig {
+  maxSockets: number
+  maxFreeSockets: number
+  freeSocketTimeout: number
+}
+
+export interface RestClientConfig {
+  timeout: TimeoutSpec
+  agent: AgentConfig
+}
+
+const NOT_FOUND = 404
+const UNAUTHORIZED = 401
+
+interface TokenSource {
+  (): Promise<string>
+}
+
+export const constantTokenSource = (token): TokenSource => async () => {
+  if (!token) {
+    throw Error('Unauthorised access')
+  }
+  return token
+}
+
+export const dynamicTokenSource = (signInService, oauthServiceName: string): TokenSource => async () => {
+  const token = await signInService.getAnonymousClientCredentialsTokens(oauthServiceName)
+  if (!token?.token) {
+    throw Error(`Error obtaining OAuth token`)
+  }
+  return token.token
+}
+
+/**
+ * Configure and return a REST api client.  This is used by the REST API facades within this module.
+ */
+export const buildRestClient = (
+  tokenSource: TokenSource,
+  apiUrl: string,
+  restApiName: string,
+  config: RestClientConfig
+) => {
+  const handleError = buildErrorHandler(restApiName)
+  const keepaliveAgent = apiUrl.startsWith('https') ? new HttpsAgent(config.agent) : new Agent(config.agent)
+
+  return {
+    async getResource(path: string, headers = {}, query = {}): Promise<any> {
+      const token = await tokenSource()
+      try {
+        const result = await superagent
+          .get(`${apiUrl}${path}`)
+          .query(query)
+          .agent(keepaliveAgent)
+          .set('Authorization', `Bearer ${token}`)
+          .set(headers)
+          .retry(2, (err, res) => {
+            if (res) {
+              if (res.status >= 300 && res.status !== NOT_FOUND && res.status !== UNAUTHORIZED) {
+                logger.warn(
+                  `Retry handler found API error with status: '${res.status}', message: '${res.error?.message}'`
```

Comment:
probably don't want to retry on any 4XX client error?

Reply:
Yes, of course! No idea why I did this.  Now retrying on res.status >= 500.